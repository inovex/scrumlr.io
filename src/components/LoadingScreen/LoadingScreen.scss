@import "constants/style";

.loading-screen {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}

[theme="dark"] {
  .loading-screen {
    background-color: $color-dark-one;
  }
}
