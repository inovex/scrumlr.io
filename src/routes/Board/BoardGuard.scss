@import "constants/style";

.board-guard {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}

.board-guard__denied-link {
  color: $color-planning-pink;
  transition: all 0.08s ease-out;

  &:hover {
    filter: brightness(1.4);
    transform: scale(1.02);
  }
}

.board-guard__info {
  font-size: $text-size--large;
}

[theme="dark"] {
  .board-guard {
    background-color: $color-dark-one;
  }

  .board-guard__info {
    color: $color-white;
  }
}
