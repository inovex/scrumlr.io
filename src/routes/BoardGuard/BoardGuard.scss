@import '../../constants/style.scss';

.board-guard__denied-link {
  background-color: $color-dark;
  font-weight: bold;
  color: $color-white;
  text-decoration: none;
  cursor: pointer;
  padding: 8px 2em;
  border-radius: 12px;
}

.board-guard__denied-link:focus,
.board-guard__denied-link:hover {
  background-color: $color-dark-gray;
}