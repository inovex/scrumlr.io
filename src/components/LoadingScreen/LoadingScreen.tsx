import "./LoadingScreen.scss";
import {LoadingIndicator} from "components/LoadingIndicator";

export const LoadingScreen = () => (
  <div className="loading-screen">
    <LoadingIndicator />
  </div>
);
