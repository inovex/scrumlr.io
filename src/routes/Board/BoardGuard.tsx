import {useEffect} from "react";
import store from "store";
import {ActionFactory} from "store/action";
import {RouteComponentProps} from "react-router";
import {ApplicationState} from "types/store";
import {useSelector} from "react-redux";
import LoadingIndicator from "components/LoadingIndicator/LoadingIndicator";
import Board from "./Board";
import "./BoardGuard.scss";

export type BoardGuardProps = RouteComponentProps<{id: string}>;

const BoardGuard = (props: BoardGuardProps) => {
  const boardStatus = useSelector((state: ApplicationState) => state.board.status);

  useEffect(() => {
    const boardId = props.match.params.id;
    store.dispatch(ActionFactory.joinBoard(boardId));

    return () => {
      store.dispatch(ActionFactory.leaveBoard());
    };
  }, [props.match.params.id]);

  if (boardStatus === "accepted" || boardStatus === "ready") {
    return <Board />;
  }
  if (boardStatus === "rejected") {
    return (
      <div className="board-guard">
        <p className="board-guard__info">You have been rejected.</p>
        <a href="/" className="board-guard__denied-link">
          Return to homepage
        </a>
      </div>
    );
  }
  return (
    <div className="board-guard">
      <LoadingIndicator />
      <p className="board-guard__info">Waiting for approval.</p>
    </div>
  );
};

export default BoardGuard;
