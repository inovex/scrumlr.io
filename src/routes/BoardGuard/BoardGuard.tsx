import * as React from 'react';
import { getFirebase } from 'react-redux-firebase';
import { User } from 'firebase';
import { RouteComponentProps } from 'react-router';
// TODO: Types are not working with most recent version of Typescript.
// TODO: Use ES6 import if typings have been adjusted.
const { Redirect } = require('react-router-dom');

import Board, { BoardProps } from '../Board';
import LoadingScreen from '../../components/LoadingScreen/LoadingScreen';

export interface BoardGuardProps extends RouteComponentProps<{ id: string }> {}

export interface BoardGuardState {
  /** If firebase is ready and user data has been loaded this flag is true. */
  ready: boolean;
  /** If the current user is authenticated this flag will be true. */
  authenticated: boolean;
  /** Refers to the unsubscribe callback of firebase auth state change event. */
  unsubscribeAuthStateChange: (() => void) | undefined;
  /** Value may change once board is deleted and reference is invalid. */
  invalidReference: boolean;
}

export class BoardGuard extends React.Component<
  BoardGuardProps,
  BoardGuardState
> {
  state: BoardGuardState = {
    ready: false,
    authenticated: false,
    unsubscribeAuthStateChange: undefined,
    invalidReference: false
  };

  componentWillMount() {
    // add callback for board deletion
    const boardId = this.props.match.params.id;
    getFirebase()
      .ref(`/boards/${boardId}/config/creatorUid`)
      .on('value', (snapshot: any) => {
        if (!snapshot.exists()) {
          this.setState({ ...this.state, invalidReference: true });
        }
      });
  }

  componentDidMount() {
    const unsubscribeAuthStateChange = getFirebase()
      .auth()
      .onAuthStateChanged((user: User) => {
        // Test if user is authenticated.
        const authenticated = user !== null;
        if (authenticated) {
          // Log user presence.
          // For details see: https://firebase.googleblog.com/2013/06/how-to-build-presence-system.html
          const userRef = getFirebase().ref(`/presence/${user.uid}`);
          const amOnline = getFirebase().ref('.info/connected');
          amOnline.on('value', (snapshot: any) => {
            if (snapshot.val()) {
              userRef.onDisconnect().remove();
              userRef.set(true);
            }
          });
        }

        // At this point firebase is ready.
        this.setState(state => ({ ...state, ready: true, authenticated }));
      });
    this.setState(state => ({ ...state, unsubscribeAuthStateChange }));
  }

  componentWillUnmount() {
    // Unsubscribe from authentication state change handler.
    // This is necessary in case the user is not authenticated and will be redirected to the login form.
    const { unsubscribeAuthStateChange } = this.state;
    if (typeof unsubscribeAuthStateChange === 'function') {
      unsubscribeAuthStateChange();
    }
  }

  render() {
    const {
      ready,
      authenticated,
      unsubscribeAuthStateChange,
      invalidReference
    } = this.state;
    const url = window.location.href;

    if (invalidReference) {
      return (
        <Redirect
          to={{
            pathname: '/new',
            state: { referrer: url }
          }}
        />
      );
    }

    if (!ready) {
      return <LoadingScreen />;
    }

    // In case user is not authenticated, redirect him to the login form, which will redirect him to the board
    // after user has been logged in successfully.
    if (!authenticated) {
      const boardId = this.props.match.params.id;
      return (
        <Redirect
          to={{
            pathname: `/join/${boardId}`,
            state: { referrer: url }
          }}
        />
      );
    }

    // If the user is logged in render the board page.
    const props: Partial<BoardProps> = {
      ...this.props,
      onSignOut: unsubscribeAuthStateChange
    };
    return <Board {...props} />;
  }
}

export default BoardGuard;
