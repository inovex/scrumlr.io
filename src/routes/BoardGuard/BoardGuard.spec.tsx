import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
// TODO: Types are not working with most recent version of Typescript.
// TODO: Use ES6 import if typings have been adjusted.
const { Redirect } = require('react-router-dom');

import BoardGuard, { BoardGuardProps, BoardGuardState } from './BoardGuard';
import LoadingScreen from '../../components/LoadingScreen/LoadingScreen';

jest.mock('!svg-inline-loader!./logo.svg', () => 'svg', {
  virtual: true
});
jest.mock('!svg-inline-loader!./logo-s.svg', () => 'svg', {
  virtual: true
});
jest.mock('react-redux-firebase', () => ({
  getFirebase: () => ({
    ref: () => ({
      on: () => jest.fn()
    })
  }),
  firebaseConnect: () => () => jest.fn(),
  helpers: {
    isLoaded: () => jest.fn()
  }
}));

describe('<BoardGuard />', () => {
  let wrapper: ShallowWrapper<BoardGuardProps, BoardGuardState>;
  let props: BoardGuardProps;

  beforeEach(() => {
    props = {
      match: {
        params: { id: '-foobar' },
        isExact: true,
        path: '/boards/:id',
        url: '/boards/-foobar'
      },
      location: {
        state: {
          referrer: 'http://example.com/some/url'
        }
      } as any,
      history: null as any
    };
  });

  it('should render loading screen by default', () => {
    wrapper = shallow(<BoardGuard {...props} />);
    expect(wrapper.find(LoadingScreen)).toHaveLength(1);
  });

  it('should render a redirect if user has not been authenticated', () => {
    wrapper = shallow(<BoardGuard {...props} />);
    wrapper.setState({ ready: true, authenticated: false });
    const redirect = wrapper.find(Redirect);
    expect(redirect).toHaveLength(1);
    expect(redirect.prop('to')).toMatchSnapshot();
  });

  it('should render a board component if user is authenticated', () => {
    wrapper = shallow(<BoardGuard {...props} />);
    wrapper.setState({ ready: true, authenticated: true });
    wrapper.update();
    expect(wrapper.find(LoadingScreen)).toHaveLength(0);
    expect(wrapper.find(Redirect)).toHaveLength(0);
  });
});
