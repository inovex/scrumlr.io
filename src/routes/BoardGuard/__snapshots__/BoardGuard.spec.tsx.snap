// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardGuard /> should redirect to homepage if board is invalid 1`] = `
Object {
  "pathname": "/new",
  "state": Object {
    "referrer": "http://localhost/",
  },
}
`;

exports[`<BoardGuard /> should render a redirect if user has not been authenticated 1`] = `
Object {
  "pathname": "/join/-foobar",
  "state": Object {
    "referrer": "http://localhost/",
  },
}
`;
